woocommerce/woocommerce
# Code Freeze CLI: Add branch command

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



Depends on https://github.com/woocommerce/woocommerce/pull/37723

This PR adds a CLI command for creating a branch on Github to be used in the code release context. It determines the next release branch to create, checks if it exists, and creates one if it doesn't exist. If the branch does exist, it asks the user if a fresh one from source should be made or do nothing.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites

1. You must have a `GITHUB_TOKEN` environment variable available in your PATH. This needs to be set to a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with repo permissions enabled.
2. Fork this repo and make sure you have a release with valid version tagged.

#### Testing the CLI

1. Run the command as is. This step assumes the release branch to be created does not yet exist in your fork.
```
pnpm run utils code-freeze branch --owner <owner> --name woocommerce
```
The CLI will determine what the next release branch is based on the latest release in your fork. Check your fork to confirm the branch now exists, is the same as `trunk` and the CLI's logs make sense.
2. Run the command with a branch supplied
```
pnpm run utils code-freeze branch --owner <owner> --name woocommerce --branch some/branch/that/does/not/exist
```
Confirm the branch now exists in your fork.
3. Rerun the first command to attempt creating a branch that now already exists.
```
pnpm run utils code-freeze branch --owner <owner> --name woocommerce
```
The CLI should ask you what you'd like to do. Try `y` and `n` to see how each one behaves. Does it work as you'd expect?
